containers/src/Data/Tree.hs

```diff
@@ -181,14 +182,46 @@ mfixTree f
 instance Traversable Tree where
     traverse f (Node x ts) = liftA2 Node (f x) (traverse (traverse f) ts)
 
+-- | Folds in preorder
 instance Foldable Tree where
-    foldMap f (Node x ts) = f x `mappend` foldMap (foldMap f) ts
+    fold = foldMap id
+    {-# INLINABLE fold #-}
+
+    foldMap f = go
+      where go (Node x ts) = f x `mappend` foldMap go ts
+    {-# INLINE foldMap #-}
+
+    foldr f z = \t -> go t z  -- Use a lambda to allow inlining with two arguments
+      where
+        go (Node x ts) = f x . foldr ((.) . go) id ts
+        -- This has been found to optimize better than the more obvious definition of
```

Comment:
this is a unjustified claim. Microbenchmarks are not enough to establish that. Given that `foldr` is marked `INLINE`, it's optimized at each use case

I'd go for simpler version if the difference in benchmarks small, as GHC optimizer is unrelieable, and perfectly should be tuned to make "normal" code run smarter. I.e. in perfect future, the simple version would optimize as well.

If the difference is big, than it's should be reported to GHC upstream.

I'm also not a fan of having `INLINE` for "performance". It's not sustainable, causes code growth (makes microbenchmarks look better, but real code worse). `INLINEABLE` should be enough for specialization.

Reply:
Forgot to mention: `foldr` doesn't need to be "fast", it meant to be lazy (e.g. for lists), and can be for `Tree` as well. 

Benchmarking something like `head (toList t)` is a good benchmark for `foldr`. It should be independent of `t` size.

Reply:
> this is a unjustified claim. Microbenchmarks are not enough to establish that. Given that foldr is marked INLINE, it's optimized at each use case

Fair point, but we have to pick one implementation, and benchmarks help us decide. If you would prefer the comment said something else I'm open to suggestions.

> I'd go for simpler version if the difference in benchmarks small

Current
```
  foldr: OK (0.36s)
    11.1 ms ± 840 μs
```
Simpler
```
  foldr: OK (0.26s)
    18.2 ms ± 1.5 ms
```

> If the difference is big, than it's should be reported to GHC upstream.

That's reasonable, I can open an issue.

> I'm also not a fan of having INLINE for "performance".

It seems necessary, `INLINABLE` doesn't always work as well. Every `foldr` in containers (`Set`, `Map`, `IntSet`, `IntMap`, `FingerTree` for `Sequence`) is `INLINE` for performance.

> Forgot to mention: foldr doesn't need to be "fast", it meant to be lazy

Why not both?

> Benchmarking something like head (toList t) is a good benchmark for foldr. 

`head (toList t)` would finish instantaneously. I don't see the point of that. But I'm happy to add/replace foldr benchmarks with other use cases. There is already `toList` and `&&` is the best I could think of for something reducing to a small value.

Reply:
What's the difference between current and simpler implementation if you don't INLINE them.

> It seems necessary, INLINABLE doesn't always work as well. Every foldr in containers (Set, Map, IntSet, IntMap, FingerTree for Sequence) is INLINE for performance.

And that is bad. Don't make things worse.

> head (toList t) would finish instantaneously.

It's a correctness test. `take n (toList infinitelySwawningTree)` should always work.
e.g. `let t = Node 'x' (repeat t)` or just `t = Node 'x' [t]`.

Reply:
@phadej, inclining a fold for a simple structure generally shouldn't increase code size much if at all. What's the problem you see with it?

Reply:
I tend to want two benchmarks for `foldr`, to reflect two use patterns:

1. `length . foldr (:) []`, for the lazy production scenario.
2. Some strict left fold with abort for other common uses. It's too early in the morning for me to come up with a nice one.

Reply:
> inlining a fold for a simple structure generally shouldn't increase code size much if at all. What's the problem you see with it?

The little streams make the big rivers.

YMMV.

Reply:
Added the `length . foldr (:) []` benchmark.

> What's the difference between current and simpler implementation if you don't INLINE them.

|`foldr (&&) True`|`INLINE`|`INLINABLE`|Nothing|
|---|---|---|---|
|Current|11.0 ms ± 546 μs|23.3 ms ± 1.6 ms|23.2 ms ± 1.5 ms|
|Simple|18.2 ms ± 1.4 ms|18.3 ms ± 1.5 ms|30.0 ms ± 1.4 ms|

|`length . foldr (:) []`|`INLINE`|`INLINABLE`|Nothing|
|---|---|---|---|
|Current|17.5 ms ± 1.7 ms|21.5 ms ± 1.6 ms|21.4 ms ± 1.8 ms|
|Simple|31.3 ms ± 1.6 ms|30.0 ms ± 2.4 ms|35.6 ms ± 2.8 ms|

I'm strongly in favor of keeping the current implementation.

> It's a correctness test.

Well then that's a test, not a benchmark. Added.

Reply:
@treeowl what do you think?
And let me know if want me to add an example of 2.

Reply:
I definitely want an example of a left fold with abort. Something like

```haskell
foldr go stop t (0 :: Int)
  where
    go x r acc
      | acc > some_constant = acc
      | otherwise = r (acc + x)
```

I think that's an important `foldr` pattern, and we want to make sure it's optimized reasonably well.

Reply:
Added.